examples_tests/60.ClusteredRendering/cull_common.glsl

```diff
@@ -0,0 +1,247 @@
+	const vec2 dirZW = unpackSnorm2x16(light.direction[1]);
+	cone.direction = vec3(dirXY.xy, dirZW.x);
+
+	// cosHalfAngle
+	// Todo(achal): I cannot handle spotlights/cone intersection against AABB
+	// if it has outerHalfAngle > 90.f, hence the `max`
+	const float cosineRange = dirZW.y;
+	cone.cosHalfAngle = max(light.outerCosineOverCosineRange * cosineRange, 1e-3f);
+
+	// baseRadius
+	const float tanOuterHalfAngle = sqrt(max(1.f - (cone.cosHalfAngle * cone.cosHalfAngle), 0.f)) / cone.cosHalfAngle;
+	cone.baseRadius = cone.height * tanOuterHalfAngle;
+
+	return cone;
+}
+
+// Todo(achal): Rename to getClusterAABB
+nbl_glsl_shapes_AABB_t getCluster(in uvec3 localClusterID, in vec3 levelMinVertex, in float voxelSideLength)
+{
+	const vec3 camPos = pc.camPosGenesisVoxelExtent.xyz;
+
+	nbl_glsl_shapes_AABB_t cluster;
+	cluster.minVx = levelMinVertex + (localClusterID * voxelSideLength) + camPos;
+	cluster.maxVx = levelMinVertex + (localClusterID + uvec3(1u)) * voxelSideLength + camPos;
+	return cluster;
+}
+
+#define PLANE_COUNT 6
+vec4[PLANE_COUNT] getAABBPlanes(in nbl_glsl_shapes_AABB_t aabb)
+{
+	vec4 planes[PLANE_COUNT];
+
+	// 157
+	vec3 p0 = vec3(aabb.maxVx.x, aabb.minVx.y, aabb.minVx.z);
+	vec3 p1 = vec3(aabb.maxVx.x, aabb.minVx.y, aabb.maxVx.z);
+	vec3 p2 = vec3(aabb.maxVx.x, aabb.maxVx.y, aabb.maxVx.z);
+	planes[0].xyz = normalize(cross(p1 - p0, p2 - p0));
+	planes[0].w = dot(planes[0].xyz, p0);
+
+	// 013
+	p0 = vec3(aabb.minVx.x, aabb.minVx.y, aabb.minVx.z);
+	p1 = vec3(aabb.maxVx.x, aabb.minVx.y, aabb.minVx.z);
+	p2 = vec3(aabb.maxVx.x, aabb.maxVx.y, aabb.minVx.z);
+	planes[1].xyz = normalize(cross(p1 - p0, p2 - p0));
+	planes[1].w = dot(planes[1].xyz, p0);
+
+	// 402
+	p0 = vec3(aabb.minVx.x, aabb.minVx.y, aabb.maxVx.z);
+	p1 = vec3(aabb.minVx.x, aabb.minVx.y, aabb.minVx.z);
+	p2 = vec3(aabb.minVx.x, aabb.maxVx.y, aabb.minVx.z);
+	planes[2].xyz = normalize(cross(p1 - p0, p2 - p0));
+	planes[2].w = dot(planes[2].xyz, p0);
+
+	// 546
+	p0 = vec3(aabb.maxVx.x, aabb.minVx.y, aabb.maxVx.z);
+	p1 = vec3(aabb.minVx.x, aabb.minVx.y, aabb.maxVx.z);
+	p2 = vec3(aabb.minVx.x, aabb.maxVx.y, aabb.maxVx.z);
+	planes[3].xyz = normalize(cross(p1 - p0, p2 - p0));
+	planes[3].w = dot(planes[3].xyz, p0);
+
+	// 451
+	p0 = vec3(aabb.minVx.x, aabb.minVx.y, aabb.maxVx.z);
+	p1 = vec3(aabb.maxVx.x, aabb.minVx.y, aabb.maxVx.z);
+	p2 = vec3(aabb.maxVx.x, aabb.minVx.y, aabb.minVx.z);
+	planes[4].xyz = normalize(cross(p1 - p0, p2 - p0));
+	planes[4].w = dot(planes[4].xyz, p0);
+
+	// 762
+	p0 = vec3(aabb.maxVx.x, aabb.maxVx.y, aabb.maxVx.z);
+	p1 = vec3(aabb.minVx.x, aabb.maxVx.y, aabb.maxVx.z);
+	p2 = vec3(aabb.minVx.x, aabb.maxVx.y, aabb.minVx.z);
+	planes[5].xyz = normalize(cross(p1 - p0, p2 - p0));
+	planes[5].w = dot(planes[5].xyz, p0);
+
+	return planes;
+}
+
+bvec3 and(in bvec3 a, in bvec3 b)
+{
+	return bvec3(a.x && b.x, a.y && b.y, a.z && b.z);
+}
+
+bvec3 or(in bvec3 a, in bvec3 b)
+{
+	return bvec3(a.x || b.x, a.y || b.y, a.z || b.z);
+}
```

Comment:
add a `nbl_glsl_` and put in the appropriate global engine header (where `nbl_glsl_dot` is)

Reply:
@nipunG314 remove those